File: elixir_ring/lib/ring_system/token.ex
defmodule RingSystem.Token do
  @moduledoc """
  Token struct that travels through ring nodes.
  """

  defstruct [
    :token_id,
    :ring_id,
    :orig_input,
    :current_val,
    :remaining_hops,
    :start_time
  ]

  @type t :: %__MODULE__{
          token_id: non_neg_integer(),
          ring_id: atom(),
          orig_input: integer(),
          current_val: integer(),
          remaining_hops: non_neg_integer(),
          start_time: integer()
        }

  def new(token_id, ring_id, value, hops) do
    %__MODULE__{
      token_id: token_id,
      ring_id: ring_id,
      orig_input: value,
      current_val: value,
      remaining_hops: hops,
      start_time: System.monotonic_time(:microsecond)
    }
  end
end

File: java_ring/src/main/java/ringsystem/Token.java
package ringsystem;

public class Token {
    public final int tokenId;
    public final RingType ringId;
    public final long origInput;
    public long currentVal;
    public int remainingHops;
    public final long startTimeNanos;

    public Token(int tokenId, RingType ringId, long origInput, int hops) {
        this.tokenId = tokenId;
        this.ringId = ringId;
        this.origInput = origInput;
        this.currentVal = origInput;
        this.remainingHops = hops;
        this.startTimeNanos = System.nanoTime();
    }
}
